File: services/catalog/src/metrics.ts
// Prometheus metrics shared across the service. Exposed at GET /metrics and
// scraped by Prometheus (see infra/prometheus/prometheus.yml).
import { collectDefaultMetrics, Counter, Histogram, Registry } from "prom-client";

export const registry = new Registry();
collectDefaultMetrics({ register: registry, prefix: "catalog_" });

export const httpRequests = new Counter({
  name: "catalog_http_requests_total",
  help: "Total HTTP requests",
  labelNames: ["method", "route", "status"],
  registers: [registry],
});

// Buckets tuned for a fast CRUD service (sub-second). p50/p95/p99 are derived
// from this histogram in Grafana.
export const httpDuration = new Histogram({
  name: "catalog_http_request_duration_seconds",
  help: "HTTP request latency in seconds",
  labelNames: ["method", "route", "status"],
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5],
  registers: [registry],
});

export const eventsEmitted = new Counter({
  name: "catalog_events_emitted_total",
  help: "Domain events emitted to Kafka",
  labelNames: ["type"],
  registers: [registry],
});

// Express middleware: times every request and records method/route/status.
export function metricsMiddleware(req: any, res: any, next: any): void {
  const end = httpDuration.startTimer();
  res.on("finish", () => {
    const route = req.route?.path ?? req.path ?? "unknown";
    const labels = { method: req.method, route, status: String(res.statusCode) };
    end(labels);
    httpRequests.inc(labels);
  });
  next();
}

File: services/catalog/app/metrics.py
"""Prometheus metrics + an ASGI middleware that times every request and labels
it by method, route template, and status. Scraped at GET /metrics."""
from __future__ import annotations

import time

from prometheus_client import Counter, Histogram
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request

HTTP_REQUESTS = Counter(
    "catalog_http_requests_total",
    "Total HTTP requests",
    ["method", "route", "status"],
)
HTTP_DURATION = Histogram(
    "catalog_http_request_duration_seconds",
    "HTTP request latency in seconds",
    ["method", "route", "status"],
    buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5),
)
EVENTS_EMITTED = Counter(
    "catalog_events_emitted_total",
    "Domain events emitted to Kafka",
    ["type"],
)


class MetricsMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        start = time.perf_counter()
        response = await call_next(request)
        # Use the matched route template (e.g. /movies/{id}) to keep cardinality
        # bounded; fall back to the raw path when no route matched.
        route = request.scope.get("route")
        path = getattr(route, "path", request.url.path)
        labels = (request.method, path, str(response.status_code))
        HTTP_DURATION.labels(*labels).observe(time.perf_counter() - start)
        HTTP_REQUESTS.labels(*labels).inc()
        return response
